torch/_higher_order_ops/associative_scan.py

```diff
@@ -40,10 +42,55 @@ def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
     return combined_leaves
 
 
+def check_args(input, combine_fn, leaves, tree, dim):
+    assert len(leaves) >= 1, "expected at least 1 input leaf"
+    assert all(
+        isinstance(x, torch.Tensor) for x in leaves
+    ), "input leaves must be a Tensor"
+    shape = leaves[0].shape
+    ndim = len(shape)
+    dim = utils.canonicalize_dim(ndim, dim)
+
+    for x in leaves[1:]:
+        assert x.shape == shape, "All input tensors must have the same shape"
+
+    out = combine_fn(
+        pytree.tree_unflatten(
+            [e[slice_along_axis(0, 1, stride=None, dim=0)] for e in leaves], tree
+        ),
+        pytree.tree_unflatten(
+            [e[slice_along_axis(0, 1, stride=None, dim=0)] for e in leaves], tree
+        ),
+    )
+    
+    out_leaves, tree_out = pytree.tree_flatten(out)
+    assert (
+        tree == tree_out
+    ), "The pytree of the output of the operator needs to match the input pytree"
+    
+    generic_scan_required = False
+    combine_fn = make_fx(combine_fn)(pytree.tree_unflatten(
```

Comment:
Do we have to run this check under dynamo? I would have thought we could call `associative_scan_op`, and inside the compiler perform this check and if it fails we use the eager implementation as a decomposition.

Reply:
Thank you for your review and your comment. Unfortunately, I don't fully understand it and I was wondering whether you could elaborate a bit more on it for me?
I was under the impression that the goal is to have a `generic_scan` (which can either run in `eager` mode or under `torch.compile`) that is more general than the optimized associative scan that invokes `tl.associative_scan`. The idea was then to determine apriori which version of scan to use, the generic scan or the non generic scan. This involves checking whether the `combine_fn` contains only pointwise operations and I didn't know about any better way to do this, but this check fails under dynamo. I am not sure how and where you envision to move this check?

Also, regarding your comment below. I just used a similar methodology as in other HOPs, such as [cond](https://github.com/pytorch/pytorch/blob/998e72cd435e73700d84aa489277866e2d12bc59/torch/_higher_order_ops/cond.py#L139-L144). I thought that the result should be that the `associative_scan` function is always compiled and that's why this if clause is placed there. Are you suggesting to remove this and expect the user to invoke `torch.compile` instead, because there could be cases where one doesn't want to have this function compiled?

Reply:
> I am not sure how and where you envision to move this check?

Inside the compiler when compiling `associative_scan_op` we can choose to fallback to the generic implementation instead. Ideally, this could be done in `torch/_inductor/decompositions.py` .

> Also, regarding your comment below. I just used a similar methodology as in other HOPs, such as [cond](https://github.com/pytorch/pytorch/blob/998e72cd435e73700d84aa489277866e2d12bc59/torch/_higher_order_ops/cond.py#L139-L144)

Looks like cond is being compiled with the eager backend there which doesn't make a great deal of sense to me tbh. 

cc @zou3519 @aakhundov  any comment on why HOPs are being compiled with eager backend instead of just running as normal eager code?

Reply:
cc @ydwu4

Reply:
> any comment on why HOPs are being compiled with eager backend instead of just running as normal eager code

We rely on dynamo to 1. check and raise error when there're any side-effects in the udf 2. lift closures as implicit inputs 3. flatten/unflatten pytree.